sycl/source/detail/context_impl.cpp

```diff
@@ -230,6 +242,121 @@ bool context_impl::isBufferLocationSupported() const {
+      InitEventsRef.erase(NewEnd, InitEventsRef.end());
+      // If there are no more events, we can mark it as fully initialized.
+      if (InitEventsRef.empty())
+        InitRef.MDeviceGlobalsFullyInitialized = true;
+      return InitEventsRef;
+    } else if (InitRef.MDeviceGlobalsFullyInitialized) {
+      // MDeviceGlobalsFullyInitialized could have been set while we were
+      // waiting on the lock and since there were no init events we are done.
+      return {};
+    }
+
+    // There were no events and it was not set as fully initialized, so this is
+    // responsible for intializing the device globals.
+    auto DeviceGlobals = InitRef.MBinImage->getDeviceGlobals();
+    std::vector<std::string> DeviceGlobalIds;
+    DeviceGlobalIds.reserve(DeviceGlobals.size());
+    for (const pi_device_binary_property &DeviceGlobal : DeviceGlobals)
+      DeviceGlobalIds.push_back(DeviceGlobal->Name);
+    std::vector<DeviceGlobalMapEntry *> DeviceGlobalEntries =
+        detail::ProgramManager::getInstance().getDeviceGlobalEntries(
+            DeviceGlobalIds,
+            /*ExcludeDeviceImageScopeDecorated=*/true);
+
+    // If there were no device globals without device_image_scope the device
+    // globals are trivially fully initialized and we can end early.
+    if (DeviceGlobalEntries.empty()) {
+      InitRef.MDeviceGlobalsFullyInitialized = true;
+      return {};
+    }
+
+    // We may have reserved too much for DeviceGlobalEntries, but now that we
+    // know number of device globals to initialize, we can use that for the
+    // list.
+    InitEventsRef.reserve(DeviceGlobalEntries.size());
+
+    // Device global map entry pointers will not die before the end of the
+    // program and the pointers will stay the same, so we do not need
+    // m_DeviceGlobalsMutex here.
+    for (DeviceGlobalMapEntry *DeviceGlobalEntry : DeviceGlobalEntries) {
+      // Get or allocate the USM memory associated with the device global.
+      DeviceGlobalUSMMem &DeviceGlobalUSM =
+          DeviceGlobalEntry->getOrAllocateDeviceGlobalUSM(QueueImpl,
+                                                          /*ZeroInit=*/true);
+
+      // If the device global still has a zero-initialization event it should be
+      // added to the initialization events list.
+      std::optional<RT::PiEvent> ZIEvent =
+          DeviceGlobalUSM.getZeroInitEvent(Plugin);
+      if (ZIEvent.has_value())
+        InitEventsRef.push_back(*ZIEvent);
+
+      // Write the pointer to the device global and store the event in the
+      // initialize events list.
+      RT::PiEvent InitEvent;
+      void *USMPtr = DeviceGlobalUSM.getPtr();
+      Plugin.call<PiApiKind::piextEnqueueDeviceVariableWrite>(
+          QueueImpl->getHandleRef(), NativePrg,
+          DeviceGlobalEntry->MUniqueId.c_str(), false, sizeof(void *), 0,
+          &USMPtr, 0, nullptr, &InitEvent);
+
+      InitEventsRef.push_back(InitEvent);
+    }
+
+    return InitEventsRef;
+  }
```

Comment:
Please consider the following:
Instead of handling a lot of events and checking their completion we could submit a queue barrier, so no new tasks start until global vars initialization finishes. Probably, we could have some corner cases where it would produce unnecessary dependencies, like
```
queue.memcpy() // Task A
queue.submit() // Task B
queue.memcpy() // Task C
```
Task C will wait for initialization of global vars for task B. But, probably, it worth code simplification and possible overhead reduction.

Reply:
The runtime knows which device binaries have kernels that use the different device_globals, so I believe it may be even more likely to run into a case where the barrier may be a little too aggressive. That is, for example a kernel A may be using a device_global G and kernel B may not, so submitting A and then B we could potentially have B wait for G to initialize. Note also that, though most device_globals are small and should be fast to initialize, we could have cases with many device_globals and/or big device_globals.

I agree, this adds some amount of complexity in the event dependency tracking, but we could consider generalizing it so we can reuse it to optimize future features and leave only one point of maintenance.